From 2b855849e68a361a8e7d00e112b03ebd26bd453b Mon Sep 17 00:00:00 2001
From: Oliver Eftevaag <oliver.eftevaag@qt.io>
Date: Wed, 24 Sep 2025 13:33:20 +0200
Subject: [PATCH 088/239] FileDialog: Comply with accessibility insights for
 Windows

Accessibility Insights for Windows is a Windows application that will
check the UI accessibility tree for any potential issues or
anti-patterns.

The tool would complain about the FileDialog having a button without a
name, in the FolderBreadcrumbBar.

The SideBar buttonDelegate would use a Button with an IconImage
contentItem, which the tool would complain about, since the Button
control then wouldn't have a name.

The FileDialogDelegate would also expose every label that it uses to
QAccessible. I'd assume users of screen readers would find it
overwhelming that every delegate has more than 4 different labels that
are all exposed to external assistive technologies. The most important
item to expose is the delegate itself, the sub-labels showing the file
size and last modified date can probably be ignored.

Pick-to: 6.8
Change-Id: Iad3fbf4a4178934214028214947506eb9d8c2651
Reviewed-by: Mitch Curtis <mitch.curtis@qt.io>
(cherry picked from commit ab83b2d2d34095bd56ca0e9e044374d314277321)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 .../qml/+Fusion/FileDialogDelegate.qml                      | 2 ++
 .../qml/+Fusion/FolderBreadcrumbBar.qml                     | 1 +
 .../quickdialogsquickimpl/qml/+Fusion/SideBar.qml           | 2 ++
 .../qml/+Imagine/FileDialogDelegate.qml                     | 2 ++
 .../qml/+Imagine/FolderBreadcrumbBar.qml                    | 1 +
 .../quickdialogsquickimpl/qml/+Imagine/SideBar.qml          | 2 ++
 .../qml/+Material/FileDialogDelegate.qml                    | 2 ++
 .../qml/+Material/FolderBreadcrumbBar.qml                   | 1 +
 .../quickdialogsquickimpl/qml/+Material/SideBar.qml         | 2 ++
 .../qml/+Universal/FileDialogDelegate.qml                   | 2 ++
 .../qml/+Universal/FolderBreadcrumbBar.qml                  | 1 +
 .../quickdialogsquickimpl/qml/+Universal/SideBar.qml        | 2 ++
 .../quickdialogsquickimpl/qml/FileDialogDelegate.qml        | 2 ++
 .../quickdialogsquickimpl/qml/FileDialogDelegateLabel.qml   | 6 ++++++
 .../quickdialogsquickimpl/qml/FolderBreadcrumbBar.qml       | 1 +
 src/quickdialogs/quickdialogsquickimpl/qml/SideBar.qml      | 2 ++
 16 files changed, 31 insertions(+)

diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/FileDialogDelegate.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/FileDialogDelegate.qml
index cb27c82034..1f6437e928 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/FileDialogDelegate.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/FileDialogDelegate.qml
@@ -40,6 +40,8 @@ DialogsQuickImpl.FileDialogDelegate {
 
     required property int fileDetailRowWidth
 
+    Accessible.name: fileName
+
     contentItem: DialogsQuickImpl.FileDialogDelegateLabel {
         delegate: control
         fileDetailRowTextColor: control.highlighted ? Fusion.highlightedText(control.palette) : control.palette.text
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/FolderBreadcrumbBar.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/FolderBreadcrumbBar.qml
index 75f1963368..40f8ddb0cd 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/FolderBreadcrumbBar.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/FolderBreadcrumbBar.qml
@@ -66,6 +66,7 @@ DialogsQuickImpl.FolderBreadcrumbBar {
         width: height
         height: Math.max(implicitHeight, control.contentItem.height)
         focusPolicy: Qt.TabFocus
+        Accessible.name: qsTr("Up")
     }
     textField: TextField {
         text: (control.dialog as DialogsQuickImpl.FileDialogImpl)?.selectedFile
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/SideBar.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/SideBar.qml
index c9dcf17f9d..6493cc47b8 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/SideBar.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Fusion/SideBar.qml
@@ -61,6 +61,8 @@ DialogsQuickImpl.SideBar {
 
         required property int index
         required property string folderName
+
+        Accessible.name: folderName
     }
 
     separatorDelegate: Item {
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/FileDialogDelegate.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/FileDialogDelegate.qml
index 54c8447f9d..be665b8fde 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/FileDialogDelegate.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/FileDialogDelegate.qml
@@ -45,6 +45,8 @@ DialogsQuickImpl.FileDialogDelegate {
 
     required property int fileDetailRowWidth
 
+    Accessible.name: fileName
+
     contentItem: DialogsQuickImpl.FileDialogDelegateLabel {
         delegate: control
         fileDetailRowTextColor: control.icon.color
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/FolderBreadcrumbBar.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/FolderBreadcrumbBar.qml
index 8becba5108..f7173f2f44 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/FolderBreadcrumbBar.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/FolderBreadcrumbBar.qml
@@ -50,6 +50,7 @@ DialogsQuickImpl.FolderBreadcrumbBar {
         icon.width: 16
         icon.height: 16
         focusPolicy: Qt.TabFocus
+        Accessible.name: qsTr("Up")
     }
     textField: TextField {
         text: (control.dialog as DialogsQuickImpl.FileDialogImpl)?.selectedFile
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/SideBar.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/SideBar.qml
index eeaaf7e80d..9884aa8d3c 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/SideBar.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Imagine/SideBar.qml
@@ -41,6 +41,8 @@ DialogsQuickImpl.SideBar {
 
         required property int index
         required property string folderName
+
+        Accessible.name: folderName
     }
 
     separatorDelegate: Item {
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Material/FileDialogDelegate.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Material/FileDialogDelegate.qml
index eab7f3f8b0..a428341cea 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Material/FileDialogDelegate.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Material/FileDialogDelegate.qml
@@ -37,6 +37,8 @@ DialogsQuickImpl.FileDialogDelegate {
 
     required property int fileDetailRowWidth
 
+    Accessible.name: fileName
+
     contentItem: DialogsQuickImpl.FileDialogDelegateLabel {
         delegate: control
         fileDetailRowTextColor: control.Material.hintTextColor
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Material/FolderBreadcrumbBar.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Material/FolderBreadcrumbBar.qml
index f049304407..58c673cc22 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Material/FolderBreadcrumbBar.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Material/FolderBreadcrumbBar.qml
@@ -64,6 +64,7 @@ DialogsQuickImpl.FolderBreadcrumbBar {
         icon.height: 16
         width: height
         focusPolicy: Qt.TabFocus
+        Accessible.name: qsTr("Up")
     }
     textField: TextField {
         text: (control.dialog as DialogsQuickImpl.FileDialogImpl)?.selectedFile
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Material/SideBar.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Material/SideBar.qml
index c6fdbd06ab..89d68ccdd6 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Material/SideBar.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Material/SideBar.qml
@@ -46,6 +46,8 @@ DialogsQuickImpl.SideBar {
 
         required property int index
         required property string folderName
+
+        Accessible.name: folderName
     }
 
     separatorDelegate: Item {
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/FileDialogDelegate.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/FileDialogDelegate.qml
index 69245500c0..9c55c6e8a6 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/FileDialogDelegate.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/FileDialogDelegate.qml
@@ -38,6 +38,8 @@ DialogsQuickImpl.FileDialogDelegate {
 
     required property int fileDetailRowWidth
 
+    Accessible.name: fileName
+
     contentItem: DialogsQuickImpl.FileDialogDelegateLabel {
         delegate: control
         fileDetailRowTextColor: control.icon.color
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/FolderBreadcrumbBar.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/FolderBreadcrumbBar.qml
index 4e730c7d00..52852f89af 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/FolderBreadcrumbBar.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/FolderBreadcrumbBar.qml
@@ -62,6 +62,7 @@ DialogsQuickImpl.FolderBreadcrumbBar {
         icon.height: 16
         width: height
         focusPolicy: Qt.TabFocus
+        Accessible.name: qsTr("Up")
     }
     textField: TextField {
         text: (control.dialog as DialogsQuickImpl.FileDialogImpl)?.selectedFile
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/SideBar.qml b/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/SideBar.qml
index 5c2ba0aabb..2ea356ee5b 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/SideBar.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/+Universal/SideBar.qml
@@ -47,6 +47,8 @@ DialogsQuickImpl.SideBar {
 
         required property int index
         required property string folderName
+
+        Accessible.name: folderName
     }
 
     separatorDelegate: Item {
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/FileDialogDelegate.qml b/src/quickdialogs/quickdialogsquickimpl/qml/FileDialogDelegate.qml
index 13a1fb127c..513d9bbcab 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/FileDialogDelegate.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/FileDialogDelegate.qml
@@ -40,6 +40,8 @@ DialogsQuickImpl.FileDialogDelegate {
 
     property int fileDetailRowWidth
 
+    Accessible.name: fileName
+
     contentItem: DialogsQuickImpl.FileDialogDelegateLabel {
         delegate: control
         fileDetailRowTextColor: control.icon.color
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/FileDialogDelegateLabel.qml b/src/quickdialogs/quickdialogsquickimpl/qml/FileDialogDelegateLabel.qml
index a28d9dea08..e285af8b28 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/FileDialogDelegateLabel.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/FileDialogDelegateLabel.qml
@@ -40,6 +40,8 @@ Item {
                 text: root.delegate.fileName
                 color: root.delegate.icon.color
                 y: (parent.height - height) / 2
+
+                Accessible.ignored: true
             }
         }
 
@@ -58,12 +60,16 @@ Item {
                 }
                 font.pixelSize: root.delegate.font.pixelSize * 0.75
                 color: root.fileDetailRowTextColor
+
+                Accessible.ignored: true
             }
             Label {
                 text: Qt.formatDateTime(root.delegate.fileModified)
                 font.pixelSize: root.delegate.font.pixelSize * 0.75
                 color: root.fileDetailRowTextColor
                 x: parent.width - width
+
+                Accessible.ignored: true
             }
         }
     }
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/FolderBreadcrumbBar.qml b/src/quickdialogs/quickdialogsquickimpl/qml/FolderBreadcrumbBar.qml
index 38af2e1bd7..2e7fbfd822 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/FolderBreadcrumbBar.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/FolderBreadcrumbBar.qml
@@ -61,6 +61,7 @@ DialogsQuickImpl.FolderBreadcrumbBar {
         icon.height: 16
         width: height
         focusPolicy: Qt.TabFocus
+        Accessible.name: qsTr("Up")
     }
     textField: TextField {
         text: (control.dialog as DialogsQuickImpl.FileDialogImpl)?.selectedFile
diff --git a/src/quickdialogs/quickdialogsquickimpl/qml/SideBar.qml b/src/quickdialogs/quickdialogsquickimpl/qml/SideBar.qml
index 231d2da0cb..5305ab1c40 100644
--- a/src/quickdialogs/quickdialogsquickimpl/qml/SideBar.qml
+++ b/src/quickdialogs/quickdialogsquickimpl/qml/SideBar.qml
@@ -51,6 +51,8 @@ DialogsQuickImpl.SideBar {
 
         required property int index
         required property string folderName
+
+        Accessible.name: folderName
     }
 
     separatorDelegate: Item {
-- 
2.51.2

